### PR TITLE
[TTAHUB-2857] Prevent closing goals if they have open objectives

### PR DESCRIPTION
**Pull request by thewatermethod**
## Description of change
- Prevent closing goals if they have open objectives
- Add objective status change to the goal cards

## How to test
- Attempt to close a goal from the RTR when objectives are open
- Note that an error message appears. Verify the link goes to the correct recipient record page.
- Close those objectives from the RTR using the inline control
- Verify that suspending an objective triggers the modal
- Verify that once all the objectives are closed or suspended, you can close a goal from the RTR
- Verify that you can't change objective status if the goal is closed

## Issue(s)
* https://jira.acf.gov/browse/TTAHUB-2857

### Every PR

- [ ] Meets issue criteria
- [ ] JIRA ticket status updated
- [ ] Code is meaningfully tested
- [ ] Meets accessibility standards (WCAG 2.1 Levels A, AA)
- [ ] API Documentation updated
- [ ] Boundary diagram updated
- [ ] Logical Data Model updated
- [ ] [Architectural Decision Records](https://adr.github.io/) written for major infrastructure decisions
- [ ] UI review complete

### Before merge to main

- [ ] OHS demo complete
- [ ] Ready to create production PR

### Production Deploy

- [ ] Staging smoke test completed

### After merge/deploy

- [ ] Update JIRA ticket status

### CODE REVIEW

File: frontend/src/hooks/useObjectiveStatusMonitor.js [hunk 1]
```diff
@@ -0,0 +1,45 @@
+import { useState, useCallback, useRef } from 'react';
+
+export default function useObjectiveStatusMonitor(objectives) {
+  const objectiveMap = useRef(
+    new Map(
+      // built to be invulnerable
```
Comment (AdamAdHocTeam) [inline frontend/src/hooks/useObjectiveStatusMonitor.js:6]: lol

File: frontend/src/hooks/useObjectiveStatusMonitor.js [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+      const objective = objectiveMap.current.get(JSON.stringify((objectiveIds || []).sort()));
+      if (!objective) {
+        return;
+      }
+
+      // update with the new status
+      objective.status = localStatus;
+
+      setAtLeastOneObjectiveIsNotCompletedOrSuspended(
+        Array.from(objectiveMap.current.values()).some((o) => o.status !== 'Complete' && o.status !== 'Suspended'),
```
Comment (AdamAdHocTeam) [inline frontend/src/hooks/useObjectiveStatusMonitor.js:32]: Should this go in a function as its used in two locations.

Comment (thewatermethod) [reply frontend/src/hooks/useObjectiveStatusMonitor.js]: sure, good idea

File: src/services/objectives.ts [hunk 3]
```diff
@@ -252,3 +290,43 @@ export async function getObjectivesByReportId(reportId) {
...
+    return false;
+  }
+
+  return true;
+}
+
+export function updateObjectiveStatusByIds(
+  objectiveIds: number[],
+  status: string,
+  closeSuspendReason = '',
```
[the remark this answers is not in the document]
Comment (AdamAdHocTeam) [reply src/services/objectives.ts:319]: disregard this is objectives... duh

**Review by AdamAdHocTeam — APPROVED** [review]
Looks good and works from the UI left some minor comments.

**Pull request merged**